codec-http2/src/main/java/io/netty/handler/codec/http2/DelegatingDecompressorFrameListener.java

```diff
@@ -97,43 +96,41 @@ public int onDataRead(ChannelHandlerContext ctx, int streamId, ByteBuf data, int
                 // not be provided with data and thus could not return how many bytes were processed. We will assume
                 // there is more data coming which will complete the decompression block. To allow for more data we
                 // return all bytes to the flow control window (so the peer can send more data).
-                decompressor.incrementDecompressedByes(compressedBytes);
-                processedBytes = compressedBytes;
-            } else {
-                try {
-                    decompressor.incrementDecompressedByes(padding);
-                    for (;;) {
-                        ByteBuf nextBuf = nextReadableBuf(channel);
-                        boolean decompressedEndOfStream = nextBuf == null && endOfStream;
-                        if (decompressedEndOfStream && channel.finish()) {
-                            nextBuf = nextReadableBuf(channel);
-                            decompressedEndOfStream = nextBuf == null;
-                        }
-
-                        decompressor.incrementDecompressedByes(buf.readableBytes());
-                        processedBytes += listener.onDataRead(ctx, streamId, buf, padding, decompressedEndOfStream);
-                        if (nextBuf == null) {
-                            break;
-                        }
-
-                        padding = 0; // Padding is only communicated once on the first iteration
-                        buf.release();
-                        buf = nextBuf;
+                decompressor.incrementDecompressedBytes(compressedBytes);
+                return compressedBytes;
+            }
+            try {
+                Http2LocalFlowController flowController = connection.local().flowController();
+                decompressor.incrementDecompressedBytes(padding);
+                for (;;) {
+                    ByteBuf nextBuf = nextReadableBuf(channel);
+                    boolean decompressedEndOfStream = nextBuf == null && endOfStream;
+                    if (decompressedEndOfStream && channel.finish()) {
+                        nextBuf = nextReadableBuf(channel);
+                        decompressedEndOfStream = nextBuf == null;
+                    }
+
+                    decompressor.incrementDecompressedBytes(buf.readableBytes());
+                    flowController.consumeBytes(stream,
```

Comment:
The `decompressor` just keeps state to be able to convert bytes the users sees (decompressed) to bytes the flow controller knows about (compressed). `ConsumedBytesConverter` wraps the flow controller so that when ever bytes are returned it uses the `decompressor` to translate from what the user knows (decompressed) to what the flow controller knows (compressed). So I think we actually want to return bytes to the flow controller here ... this is effectively what is done when bytes are returned by this method (done by `DefaultHttp2ConnectionDecoder`) however bcz there may be multiple frames decompressed it makes sense to return the bytes immediately instead of waiting for the method to return.